### PR TITLE
Handle errors using Rails.error

### PR DESCRIPTION
**Pull request by johngallagher**
Jira ticket: BIG-8954

## What

Detail error handling from Rails 7

## Why

To help folk decide how to use it and provide some basic docs around usage.

### CODE REVIEW

**Review by cernivec — DISMISSED** [review]

**Comment by RenatoCoelho-BiggerPockets** [comment]
Strange how PRs from the patterns repo don't automatically add jira links to: Jira ticket: BIG-8954

**Review by karlentwistle — DISMISSED** [review]

**Review by RenatoCoelho-BiggerPockets — APPROVED** [review]

**Review by bockets — APPROVED** [review]
It's a bit wordy, but I agree with the pattern. The benefits are mostly obvious to me. Could we maybe trim this down?

File: _patterns/prefer-rescuing-using-active-support-error-handler.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  def http_client
+    Faraday.new { |f| f.response :raise_error }
+  end
+end
+```
+
+This has many disadvantages:
+
+1. **Messy** There are three lines of noise to wade through.
+2. **Cognitively heavy** Should we log an error? Or just a warning? Should we send to Sentry or not? Often this overhead means developers fall into sloppy practices, don't think through error handling and it's done inconsistently or just plain wrong.
```
Comment (karlentwistle) [inline _patterns/prefer-rescuing-using-active-support-error-handler.md:32]:
```suggestion
2. **Cognitively heavy** Should we log an error? Or just a warning? Should we send to Sentry or not? Often this overhead means developers fall into sloppy practices, don't think through error handling and it's done inconsistently.
```

I think "or just plain wrong" is a bit of a bummer / aggressive and also highly subjective. Could we just ✂️ this 😅

**Review by BPMattBauer — DISMISSED** [review]

**Review by DeadKennedyx — DISMISSED** [review]

**Comment by johngallagher** [comment]
> Strange how PRs from the patterns repo don't automatically add jira links to: Jira ticket: [BIG-8954](https://biggerpockets.atlassian.net/browse/BIG-8954)

Great point - fixed! It's set up in [autolink settings](https://github.com/BiggerPockets/patterns/settings/key_links)

File: _patterns/prefer-rescuing-using-active-support-error-handler.md (hunk 1 above)
Comment (johngallagher) [reply _patterns/prefer-rescuing-using-active-support-error-handler.md:32]: Great point @karlentwistle  - think I fell into being too ranty here. Thanks. :)

**Comment by johngallagher** [comment]
> It's a bit wordy, but I agree with the pattern. The benefits are mostly obvious to me. Could we maybe trim this down?

I love it! Thanks for the feedback. I'll put my brutal editing hat on and trim it down.

**Review by BPMattBauer — APPROVED** [review]

File: _patterns/prefer-rescuing-using-active-support-error-handler.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  def http_client
+    Faraday.new { |f| f.response :raise_error }
+  end
+end
+```
+
+This has many disadvantages:
+
+1. **Cognitively heavy** Every time we rescue, we need to decide what to do.
```
Comment (RenatoCoelho-BiggerPockets) [inline _patterns/prefer-rescuing-using-active-support-error-handler.md:31]: Its pet peeve but could you separate the bold text from the description?

**Cognitively heavy** - Every time we rescue, we need to decide what to do.
**Cognitively heavy**, Every time we rescue, we need to decide what to do.
**Cognitively heavy**. Every time we rescue, we need to decide what to do.

**Review by RenatoCoelho-BiggerPockets — APPROVED** [review]

File: _patterns/prefer-rescuing-using-active-support-error-handler.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+2. **Opaque intent** No abstractions mean we're not communicating effectively with humans.
+3. **Tightly coupled, duplicate code** When we change how we log, these changes will ripple out.
+4. **Missing structured data in logs** We can build context into the error and send this to the logs.
```
Comment (bockets) [inline _patterns/prefer-rescuing-using-active-support-error-handler.md:34]:
```suggestion
1. **Opaque intent** No abstractions mean we're not communicating effectively with humans.
1. **Tightly coupled, duplicate code** When we change how we log, these changes will ripple out.
1. **Missing structured data in logs** We can build context into the error and send this to the logs.
```

**Review by bockets — COMMENTED** [review]
little nit

**Review by bockets — APPROVED** [review]

**Review by kxhitiz — APPROVED** [review]

**Review by sogamoso — APPROVED** [review]

File: _patterns/prefer-rescuing-using-active-support-error-handler.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+Three options:
+
+* `#handle` - most common. Rescues the exception and swallows it. `warning` severity.
+* `#record` - rescues, but then reraises exception. `error` severity.
+* `#report` - use with legacy code with complex error handling.
+
+```mermaid
+graph TD
+    A[Exception in code] --> C{1. Is the exception</br>already rescued</br>by other code?}
```
Comment (johngallagher) [inline _patterns/prefer-rescuing-using-active-support-error-handler.md:82]: Make this more of a question about the existing code being complex - and give an example.

File: _patterns/prefer-rescuing-using-active-support-error-handler.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+## Which method do I use?
+
+Three options:
+
+* `#handle` - most common. Rescues the exception and swallows it. `warning` severity.
+* `#record` - rescues, but then reraises exception. `error` severity.
+* `#report` - use with legacy code with complex error handling.
+
+```mermaid
```
[the remark this answers is not in the document]
Comment (bockets) [reply _patterns/prefer-rescuing-using-active-support-error-handler.md:80]: that's the markup that makes graphs

**Review by Hafiz-ahmed-dev — APPROVED** [review]

**Comment by bockets** [comment]
Does this work with sidekiq?

**Review by DeadKennedyx — APPROVED** [review]

File: _patterns/prefer-rescuing-using-active-support-error-handler.md (hunk 2 above)
Comment (johngallagher) [reply _patterns/prefer-rescuing-using-active-support-error-handler.md]: 8b171c1

File: _patterns/prefer-rescuing-using-active-support-error-handler.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (johngallagher) [reply _patterns/prefer-rescuing-using-active-support-error-handler.md:80]: https://mermaid-js.github.io/ @RajaaBoulassouak

**Comment by johngallagher** [comment]
> Does this work with sidekiq?

Great question - let's give it a try when we add another instance of this into the codebase and find out!

File: _patterns/prefer-rescuing-using-active-support-error-handler.md (hunk 3 above)
Comment (johngallagher) [reply _patterns/prefer-rescuing-using-active-support-error-handler.md]: 81a6dd8
